orttraining/orttraining/training_ops/cuda/tensor/gather_elements_grad.cc

```diff
@@ -0,0 +1,135 @@
+                    const int64_t indices_size,
+                    TArray<int64_t>& buffer_indices_dims,
+                    TArray<fast_divmod>& fdm_indices_strides,
+                    const int axis) const {
+    T* output_data = output_tensor->template MutableData<T>();
+    const T* update_data = updates_tensor->template Data<T>();
+    typedef typename ToCudaType<T>::MappedType CudaT;
+
+    MLDataType Tin_type = indices_tensor->DataType();
+    if (utils::IsPrimitiveDataType<int32_t>(Tin_type)) {
+      const int32_t* indices_data = indices_tensor->template Data<int32_t>();
+      return GatherElementsGradImpl(
+          rank,
+          buffer_input_dims,
+          buffer_input_strides,
+          indices_data,
+          indices_size,
+          buffer_indices_dims,
+          fdm_indices_strides,
+          reinterpret_cast<const CudaT*>(update_data),
+          axis,
+          reinterpret_cast<CudaT*>(output_data));
+    } else if (utils::IsPrimitiveDataType<int64_t>(Tin_type)) {
+      const int64_t* indices_data = indices_tensor->template Data<int64_t>();
+      return GatherElementsGradImpl(
+          rank,
+          buffer_input_dims,
+          buffer_input_strides,
+          indices_data,
+          indices_size,
+          buffer_indices_dims,
+          fdm_indices_strides,
+          reinterpret_cast<const CudaT*>(update_data),
+          axis,
+          reinterpret_cast<CudaT*>(output_data));
+    }
+
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED, "Type for Tin is not supported yet in ScatterElements.");
+  }
+};
+
+Status GatherElementsGrad::ComputeInternal(OpKernelContext* context) const {
+  const Tensor* shape = context->Input<Tensor>(0);
+  const TensorShape data_shape(shape->template Data<int64_t>(), shape->Shape().Size());
+
+  const int axis = static_cast<int>(HandleNegativeAxis(axis_, data_shape.NumDimensions()));
+
+  const auto* indices_tensor = context->Input<Tensor>(1);
+  const auto* updates_tensor = context->Input<Tensor>(2);
+
+  const auto& indices_dims = indices_tensor->Shape().GetDims();
+  const int64_t indices_size = indices_tensor->Shape().Size();
+  const auto& updates_dims = updates_tensor->Shape().GetDims();
+  if (indices_dims.size() != updates_dims.size()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Indices and updates must have the same rank");
+  }
+
+  for (size_t i = 0; i < indices_dims.size(); ++i) {
+    if (indices_dims[i] != updates_dims[i]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices vs updates dimensions differs at position=", i,
+                             " ", indices_dims[i], " vs ", updates_dims[i]);
+    }
+  }
+
+  // According to the spec the rank of ind/upd shall be the same as input(data)
+  // and we also want to make sure that the dimensions of the of the ind/upd do not
+  // exceed that of the input
+  const auto& input_dims = data_shape.GetDims();
```

Comment:
the name "input" is confusing here, since it's actually op output. 

update - > dY
output_tensor -> dX

Reply:
you are right.